### PR TITLE
[Python][Dev] Dynamically generate the Connection wrapper methods

### PR DESCRIPTION
**Pull request by Tishj**
This was becoming increasingly hard to maintain, so I have decided to generate the methods instead.

Every method defined on `DuckDBPyConnection` now lives in `tools/pythonpkg/scripts/connection_methods.json`

They look like this:
```json
	{
		"name": "register_filesystem",
		"function": "RegisterFilesystem",
		"docs": "Register a fsspec compliant filesystem",
		"args": [
			{
				"name": "filesystem",
				"type": "str"
			}
		],
		"return": "None"
	}
```

From this, we generate sections of code inside of:
`tools/pythonpkg/duckdb/__init__.py`
`tools/pythonpkg/duckdb-stubs/__init__.pyi` (`DuckDBPyConnection`)
`tools/pythonpkg/duckdb-stubs/__init__.pyi` (`duckdb`)
`tools/pythonpkg/src/pyconnection.cpp` (`InitializeConnectionMethods`)

From now on, when adjusting or creating a new connection method, the `connection_methods.json` file should be updated.

## Old Process
- Create a C++ function in `pyconnection.hpp` and `pyconnection.cpp`
- Add a pybind definition for it inside `pyconnection.cpp`
- Create a wrapper C++ function in `connection_wrapper.hpp` and `connection_wrapper.cpp`
- Add a pybind definition for it inside `duckdb_python.cpp`
- Add it to `duckdb-stubs/__init__.pyi` for both the DuckDBPyConnection method and the corresponding wrapper method.
- Add the wrapper connection method to `duckdb/__init__.py` (expose the symbol from `duckdb.duckdb`)

## New Process
- Create a C++ function in `pyconnection.hpp` and `pyconnection.cpp`
- Update the `connection_methods.json`
- Run `make generate-files`

Some issues encountered:
- `values` didn't take the optional connection as a keyword argument, now it does
- `arrow` and `df` are overloaded twice, once for `from_arrow/df` and another for `to_arrow/df`.
We now define these in C++ because it's easier (pybind11 handles the overloading on the Python side).

### CODE REVIEW

**Comment by Tishj** [comment]
> Great idea to generate this - maybe we can instead of generating the function objects during `import duckdb`, have a separate step that generates the code and puts it into `init.py`, so that we don't need to execute complex code during the import?

While I agree, my source of truth for generating this is the definitions created by pybind11 for the connection methods.
Which means that to generate the `__init__.py` I would need to reference the `__init__.py`, which seems problematic

So I decided to take a step back and generate the cpp code for the DuckDBPyConnection methods that are shared, for which I am writing this gigantic json file that contains the information for both the stubs and the `m.def(...)` code, but that seems like it's going to take a bit of grunt work:
```json
	{
		"name": "create_function",
		"function": "RegisterScalarUDF",
		"docs": "Create a DuckDB function out of the passing in Python function so it can be used in queries",
		"args": [
			{
				"name": "name",
				"type": "str"
			},
			{
				"name": "function",
				"type": "function"
			},
			{
				"name": "parameters",
				"type": "Optional[List[DuckDBPyType]]",
				"default": "None"
			},
			{
				"name": "return_type",
				"type": "Optional[DuckDBPyType]",
				"default": "None"
			}
		],
		"kwargs": [
			{
				"name": "type",
				"type": "Optional[PythonUDFType]",
				"default": "PythonUDFType.NATIVE"
			},
			{
				"name": "null_handling",
				"type": "Optional[FunctionNullHandling]",
				"default": "FunctionNullHandling.DEFAULT"
			},
			{
				"name": "exception_handling",
				"type": "Optional[PythonExceptionHandling]",
				"default": "PythonExceptionHandling.DEFAULT"
			},
			{
				"name": "side_effects",
				"type": "bool",
				"default": "False"
			}
		],
		"return": "DuckDBPyConnection"
	},
```

If everything goes according to plan however, I can write a script that generates all of these things from this json file:
- stubs (connection)
- stubs (connection wrapper)
- init.py (connection)
- init.py (connection wrapper)
- cpp code (connection)

**Comment by Tishj** [comment]
It's working how I envisioned it, always a good feeling :smile:
I just need to add another script that invokes the 4 scripts for convenience, and add it to the format check job

File: tools/pythonpkg/duckdb-stubs/__init__.pyi [hunk 1]
```diff
@@ -253,147 +253,84 @@ def FunctionExpression(function: str, *cols: Expression) -> Expression: ...
...
+    def view(self, view_name: str) -> DuckDBPyRelation: ...
+    def values(self, values: List[Any]) -> DuckDBPyRelation: ...
+    def table_function(self, name: str, parameters: object = None) -> DuckDBPyRelation: ...
+    def read_json(self, name: str, **kwargs) -> DuckDBPyRelation: ...
+    def extract_statements(self, query: str) -> List[Statement]: ...
+    def sql(self, query: str, **kwargs) -> DuckDBPyRelation: ...
+    def query(self, query: str, **kwargs) -> DuckDBPyRelation: ...
+    def from_query(self, query: str, **kwargs) -> DuckDBPyRelation: ...
+    def read_csv(self, path_or_buffer: Union[str, StringIO, TextIOBase], **kwargs) -> DuckDBPyRelation: ...
+    def from_csv_auto(self, path_or_buffer: Union[str, StringIO, TextIOBase], **kwargs) -> DuckDBPyRelation: ...
```
Comment (Mytherin) [inline tools/pythonpkg/duckdb-stubs/__init__.pyi:321]: Can we expand the keyword arguments here (using `*, arg1=...`), similar to `read_parquet` in the same doc (see maybe also the old [numpy stubs](https://github.com/numpy/numpy-stubs/blob/master/numpy-stubs/__init__.pyi)).

**Comment by Mytherin** [comment]
Thanks! LGTM now

**Pull request merged**